src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs

```diff
@@ -164,7 +167,20 @@ internal async Task<CompletionParams> GetProvisionalCompletionParamsAsync(Comple
                 TextDocument = new TextDocumentIdentifier() { Uri = previousCharacterProjection.Uri }
             };
 
-            return provisionalCompletionParams;
+            result = await _requestInvoker.RequestServerAsync<CompletionParams, SumType<CompletionItem[], CompletionList>?>(
+                Methods.TextDocumentCompletionName,
+                LanguageServerKind.CSharp,
+                provisionalCompletionParams,
+                cancellationToken).ConfigureAwait(true);
+
+            // We have now obtained the necessary completion items. We no longer need the provisional change. Revert.
+            var removeProvisionalDot = new TextChange(
+                TextSpan.FromBounds(previousCharacterProjection.PositionIndex, previousCharacterProjection.PositionIndex + 1),
+                string.Empty);
+
+            trackingDocumentManager.UpdateVirtualDocument<CSharpVirtualDocument>(documentSnapshot.Uri, new[] { removeProvisionalDot }, previousCharacterProjection.HostDocumentVersion);
```

Comment:
This is the key change in this PR. Note: I explicitly call `ConfigureAwait(true)` in the above request which guarantees that we would still be on the main thread at this point.